### PR TITLE
Preprocessor: Use eval() for good

### PR DESCRIPTION
**Pull request by kripken**
This replaces our custom preprocessor code with `eval()`. This allows arbitrary conditions in our `#if`s, e.g. the `__EMSCRIPTEN_HAS_nodefs__` things are now
```
#if LibraryManager.has('library_workerfs.js')
```
where that function is a new utility that checks if a library was included.

This patch also switches to things we've wanted to do, like
```
#if FULL_ES2 || LEGACY_GL_EMULATION
```
where in the past we had to create a new global for them.

We now error if we do if we ifdef on something nonexistent (the eval call throws), which noticed that we had a bunch of `NO_DYNAMIC_EXECUTION` code that could be removed (we removed that option a while ago).

### CODE REVIEW

File: src/support.js [hunk 1]
```diff
@@ -603,15 +603,15 @@ Module['registerFunctions'] = registerFunctions;
 #endif // RELOCATABLE
 #endif // EMULATED_FUNCTION_POINTERS
 
-#if WASM_BACKEND_WITH_RESERVED_FUNCTION_POINTERS
+#if WASM_BACKEND && RESERVED_FUNCTION_POINTERS
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/support.js]: Thanks, fixed

**Comment by kripken** [comment]
I undid the embind changes regarding `NO_DYNAMIC_EXECUTION` - I realized those were wrong. We will need a different solution there, I think there's a bug in the usage of that flag.

**Comment by kripken** [comment]
Turns out #7653 will fix the `NO_DYNAMIC_EXECUTION` stuff which was an unrelated embind bug. So we just need to wait on that.

**Comment by kripken** [comment]
Actually, this blocks #7612, so let's land it with a small workaround for that embind issue, that we can remove soon.

**Comment by wffurr** [comment]
Thanks for this - I was considering making #if on a non-existent setting an error as well.  Happy to see you have already done so.

**Pull request merged**